### PR TITLE
8263707: C1 RangeCheckEliminator support constant array and NewMultiArray

### PR DESCRIPTION
**Pull request by huishi-hs**
…rray

C1 misses range check elimination opportunities for constant and NewMultiArray with fixed length now. This patch adds constant length node for load/store Indexed node when array is constant array or allocated with NewMultiArray  and its first dimension length is constant.

Tested on linux x64 release/fastdebug with tier1 and tier2.

Regards
Hui

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263707](https://bugs.openjdk.java.net/browse/JDK-8263707): C1 RangeCheckEliminator support constant array and NewMultiArray

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to c7d28819568042f47b85ff28150997d9632ca99c
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3041/head:pull/3041` \
`$ git checkout pull/3041`

Update a local copy of the PR: \
`$ git checkout pull/3041` \
`$ git pull https://git.openjdk.java.net/jdk pull/3041/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3041`

View PR using the GUI difftool: \
`$ git pr show -t 3041`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3041.diff">https://git.openjdk.java.net/jdk/pull/3041.diff</a>

</details>

### CODE REVIEW

**Comment by huishi-hs** [comment]
Could someone help review this PR?

**Comment by neliasso** [comment]
Do you have any test that exercises this code path?

**Comment by huishi-hs** [comment]
> Do you have any test that exercises this code path?

I have an example case in JBS https://bugs.openjdk.java.net/secure/attachment/93668/range_check.java.  Do you suggest adding a test case to check if RC is removed as expected?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I've also executed some internal testing and it all passed.

**Comment by neliasso** [comment]
> > Do you have any test that exercises this code path?
> 
> @neliasso
> 
> I have an example case in JBS https://bugs.openjdk.java.net/secure/attachment/93668/range_check.java. Do you suggest adding a test case to check if RC is removed as expected?

Yes - please do that. Look at the tests in open/test/hotspot/jtreg/compiler/c1 (for example RangeCheckVerificationOfIR.java).

File: src/hotspot/share/c1/c1_Canonicalizer.cpp [hunk 1]
```diff
@@ -242,10 +242,16 @@ void Canonicalizer::do_ArrayLength    (ArrayLength*     x) {
     // with same value Otherwise a Constant is live over multiple
     // blocks without being registered in a state array.
     Constant* length;
+    NewMultiArray* nma;
     if (na->length() != NULL &&
         (length = na->length()->as_Constant()) != NULL) {
       assert(length->type()->as_IntConstant() != NULL, "array length must be integer");
       set_constant(length->type()->as_IntConstant()->value());
+    } else if ((nma = x->array()->as_NewMultiArray()) != NULL) {
+      if ((length = nma->dims()->at(0)->as_Constant()) != NULL) {
```
[the remark this answers is not in the document]
Comment (huishi-hs) [reply src/hotspot/share/c1/c1_Canonicalizer.cpp]: ifs are merged in updated commit

**Comment by huishi-hs** [comment]
@neliasso

Test is added by checking if rangchecks are eliminated in given cases. -XX:+TraceRangeCheckElimination is added when launch C1 compilation, OutputAnalyzer is used to count how many RCs are eliminated. This test can only run with debug build.

RangeCheckVerificationOfIR.java constructs all kinds of CFG with loops expect pass assertions in C1 code. Actually it doesn't dump IR and check.

**Comment by huishi-hs** [comment]
May I proceed to sponsor state? @neliasso

**Review by neliasso — APPROVED** [review]
Looks good.

**Comment by neliasso** [comment]
> May I proceed to sponsor state? @neliasso

Yes, please do that. I will sponsor your PR.

**Comment by huishi-hs** [comment]
> > May I proceed to sponsor state? @neliasso
> 
> Yes, please do that. I will sponsor your PR.

Thanks for your help!

**Pull request closed without merging**